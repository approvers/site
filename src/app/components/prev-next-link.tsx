import { Button, Flex, Spacer } from "@chakra-ui/react";
import { faArrowLeft, faArrowRight } from "@fortawesome/free-solid-svg-icons";
import { FontAwesomeIcon } from "@fortawesome/react-fontawesome";
import NextLink from "next/link";

export type PrevNextLinkProps = {
  prevLinkHref: string | null;
  nextLinkHref: string | null;
};

export const PrevNextLink = ({ prevLinkHref, nextLinkHref }: PrevNextLinkProps): JSX.Element => (
  <Flex as="nav" mt={4} mb={4} color="highlighted">
    {prevLinkHref !== null && (
      <Button as={NextLink} href={prevLinkHref} leftIcon={<FontAwesomeIcon icon={faArrowLeft} />}>
        前
      </Button>
    )}
    <Spacer />
    {nextLinkHref !== null && (
      <Button as={NextLink} href={nextLinkHref} rightIcon={<FontAwesomeIcon icon={faArrowRight} />}>
        次
      </Button>
    )}
  </Flex>
);
