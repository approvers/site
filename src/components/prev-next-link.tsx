import { Button, Flex, Spacer } from "@chakra-ui/react";
import { faArrowLeft } from "@fortawesome/free-solid-svg-icons/faArrowLeft";
import { faArrowRight } from "@fortawesome/free-solid-svg-icons/faArrowRight";
import { FontAwesomeIcon } from "@fortawesome/react-fontawesome";
import { Link } from "gatsby";
import React from "react";

export type PrevNextLinkProps = {
  prevLinkHref: string | null;
  nextLinkHref: string | null;
};

export const PrevNextLink = ({ prevLinkHref, nextLinkHref }: PrevNextLinkProps): JSX.Element => (
  <Flex as="nav" mt={4} mb={4} color="highlighted">
    {prevLinkHref !== null && (
      <Button asChild>
        <Link to={prevLinkHref}>
          <FontAwesomeIcon icon={faArrowLeft} /> 前
        </Link>
      </Button>
    )}
    <Spacer />
    {nextLinkHref !== null && (
      <Button asChild>
        <Link to={nextLinkHref}>
          次 <FontAwesomeIcon icon={faArrowRight} />
        </Link>
      </Button>
    )}
  </Flex>
);
