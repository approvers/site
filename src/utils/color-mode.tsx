"use client";

import type { IconButtonProps, SpanProps } from "@chakra-ui/react";
import { ClientOnly, IconButton, Skeleton, Span } from "@chakra-ui/react";
import { ThemeProvider, useTheme } from "next-themes";
import type { ThemeProviderProps } from "next-themes";
import * as React from "react";
import { LuMoon, LuSun } from "react-icons/lu";

export interface ColorModeProviderProps extends ThemeProviderProps {}

export function ColorModeProvider(props: ColorModeProviderProps) {
  return <ThemeProvider attribute="class" disableTransitionOnChange {...props} />;
}

export type ColorMode = "light" | "dark";

export interface UseColorModeReturn {
  colorMode: ColorMode;
  setColorMode: (colorMode: ColorMode) => void;
  toggleColorMode: () => void;
}

export function useColorMode(): UseColorModeReturn {
  const { resolvedTheme, setTheme, forcedTheme } = useTheme();
  const colorMode = forcedTheme || resolvedTheme;
  const toggleColorMode = () => {
    setTheme(resolvedTheme === "dark" ? "light" : "dark");
  };
  return {
    colorMode: colorMode as ColorMode,
    setColorMode: setTheme,
    toggleColorMode,
  };
}

export function useColorModeValue<T>(light: T, dark: T) {
  const { colorMode } = useColorMode();
  return colorMode === "dark" ? dark : light;
}

export function ColorModeIcon() {
  const { colorMode } = useColorMode();
  return colorMode === "dark" ? <LuMoon /> : <LuSun />;
}

export function OppositeColorModeIcon() {
  const { colorMode } = useColorMode();
  return colorMode === "dark" ? <LuSun /> : <LuMoon />;
}

interface ColorModeButtonProps extends Omit<IconButtonProps, "aria-label"> {}

export const ColorModeButton = React.forwardRef<HTMLButtonElement, ColorModeButtonProps>(
  function ColorModeButton(props, ref) {
    const { toggleColorMode } = useColorMode();
    return (
      <ClientOnly fallback={<Skeleton boxSize="9" />}>
        <IconButton
          ref={ref}
          aria-label="Toggle color mode"
          onClick={toggleColorMode}
          size="sm"
          variant="ghost"
          {...props}
          css={{
            _icon: {
              width: "5",
              height: "5",
            },
          }}
        >
          <ColorModeIcon />
        </IconButton>
      </ClientOnly>
    );
  },
);

export const LightMode = React.forwardRef<HTMLSpanElement, SpanProps>(
  function LightMode(props, ref) {
    return (
      <Span
        className="chakra-theme light"
        ref={ref}
        display="contents"
        color="fg"
        colorPalette="gray"
        colorScheme="light"
        {...props}
      />
    );
  },
);

export const DarkMode = React.forwardRef<HTMLSpanElement, SpanProps>(function DarkMode(props, ref) {
  return (
    <Span
      className="chakra-theme dark"
      ref={ref}
      display="contents"
      color="fg"
      colorPalette="gray"
      colorScheme="dark"
      {...props}
    />
  );
});
