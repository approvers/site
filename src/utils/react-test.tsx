import reactTestRenderer, { ReactTestRenderer, TestRendererOptions } from "react-test-renderer";
import { ReactElement } from "react";

export const render = (
  ui: ReactElement,
  options?: TestRendererOptions | undefined,
): ReactTestRenderer => reactTestRenderer.create(ui, options);
