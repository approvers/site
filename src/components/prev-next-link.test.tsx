import { expect, it } from "vitest";
import { PrevNextLink } from "./prev-next-link";
import { render } from "../utils/react-test";

it("renders both prev and next correctly", () => {
  const tree = render(
    <PrevNextLink
      prevLinkHref="https://example.com/prev"
      nextLinkHref="https://example.com/next"
    />,
  );
  expect(tree).toMatchSnapshot();
});

it("renders only prev correctly", () => {
  const tree = render(<PrevNextLink prevLinkHref="https://example.com/prev" nextLinkHref={null} />);
  expect(tree).toMatchSnapshot();
});

it("renders only next correctly", () => {
  const tree = render(<PrevNextLink prevLinkHref={null} nextLinkHref="https://example.com/next" />);
  expect(tree).toMatchSnapshot();
});
