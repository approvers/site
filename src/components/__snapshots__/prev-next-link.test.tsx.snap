// Vitest Snapshot v1

exports[`renders both prev and next correctly 1`] = `
<nav
  className="css-10z1pgu"
>
  <a
    className="chakra-button css-4xx2wk"
    href="https://example.com/prev"
    onClick={[Function]}
    onMouseEnter={[Function]}
    onTouchStart={[Function]}
  >
    <span
      className="chakra-button__icon css-1wh2kri"
    >
      <svg
        aria-hidden={true}
        className="svg-inline--fa fa-arrow-left "
        data-icon="arrow-left"
        data-prefix="fas"
        focusable={false}
        role="img"
        style={{}}
        viewBox="0 0 448 512"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M9.4 233.4c-12.5 12.5-12.5 32.8 0 45.3l160 160c12.5 12.5 32.8 12.5 45.3 0s12.5-32.8 0-45.3L109.2 288 416 288c17.7 0 32-14.3 32-32s-14.3-32-32-32l-306.7 0L214.6 118.6c12.5-12.5 12.5-32.8 0-45.3s-32.8-12.5-45.3 0l-160 160z"
          fill="currentColor"
          style={{}}
        />
      </svg>
    </span>
    前
  </a>
  <div
    className="css-17xejub"
  />
  <a
    className="chakra-button css-4xx2wk"
    href="https://example.com/next"
    onClick={[Function]}
    onMouseEnter={[Function]}
    onTouchStart={[Function]}
  >
    次
    <span
      className="chakra-button__icon css-1hzyiq5"
    >
      <svg
        aria-hidden={true}
        className="svg-inline--fa fa-arrow-right "
        data-icon="arrow-right"
        data-prefix="fas"
        focusable={false}
        role="img"
        style={{}}
        viewBox="0 0 448 512"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M438.6 278.6c12.5-12.5 12.5-32.8 0-45.3l-160-160c-12.5-12.5-32.8-12.5-45.3 0s-12.5 32.8 0 45.3L338.8 224 32 224c-17.7 0-32 14.3-32 32s14.3 32 32 32l306.7 0L233.4 393.4c-12.5 12.5-12.5 32.8 0 45.3s32.8 12.5 45.3 0l160-160z"
          fill="currentColor"
          style={{}}
        />
      </svg>
    </span>
  </a>
</nav>
`;

exports[`renders only next correctly 1`] = `
<nav
  className="css-10z1pgu"
>
  <div
    className="css-17xejub"
  />
  <a
    className="chakra-button css-4xx2wk"
    href="https://example.com/next"
    onClick={[Function]}
    onMouseEnter={[Function]}
    onTouchStart={[Function]}
  >
    次
    <span
      className="chakra-button__icon css-1hzyiq5"
    >
      <svg
        aria-hidden={true}
        className="svg-inline--fa fa-arrow-right "
        data-icon="arrow-right"
        data-prefix="fas"
        focusable={false}
        role="img"
        style={{}}
        viewBox="0 0 448 512"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M438.6 278.6c12.5-12.5 12.5-32.8 0-45.3l-160-160c-12.5-12.5-32.8-12.5-45.3 0s-12.5 32.8 0 45.3L338.8 224 32 224c-17.7 0-32 14.3-32 32s14.3 32 32 32l306.7 0L233.4 393.4c-12.5 12.5-12.5 32.8 0 45.3s32.8 12.5 45.3 0l160-160z"
          fill="currentColor"
          style={{}}
        />
      </svg>
    </span>
  </a>
</nav>
`;

exports[`renders only prev correctly 1`] = `
<nav
  className="css-10z1pgu"
>
  <a
    className="chakra-button css-4xx2wk"
    href="https://example.com/prev"
    onClick={[Function]}
    onMouseEnter={[Function]}
    onTouchStart={[Function]}
  >
    <span
      className="chakra-button__icon css-1wh2kri"
    >
      <svg
        aria-hidden={true}
        className="svg-inline--fa fa-arrow-left "
        data-icon="arrow-left"
        data-prefix="fas"
        focusable={false}
        role="img"
        style={{}}
        viewBox="0 0 448 512"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M9.4 233.4c-12.5 12.5-12.5 32.8 0 45.3l160 160c12.5 12.5 32.8 12.5 45.3 0s12.5-32.8 0-45.3L109.2 288 416 288c17.7 0 32-14.3 32-32s-14.3-32-32-32l-306.7 0L214.6 118.6c12.5-12.5 12.5-32.8 0-45.3s-32.8-12.5-45.3 0l-160 160z"
          fill="currentColor"
          style={{}}
        />
      </svg>
    </span>
    前
  </a>
  <div
    className="css-17xejub"
  />
</nav>
`;
